### ISSUE TITLE
PulseAudio Volume Widget

### ISSUE THREAD

**Issue by IvanMalison** [taffybar/taffybar#428]
We could use this dbus interface

https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/Developer/Clients/DBus/

**Comment by u11gh** [comment]
Started working on it [here](https://github.com/u11gh/taffybar/tree/fix%23428). (Though at the moment it's more like a "proof of concept".)

**Comment by IvanMalison** [comment]
@u11gh I think it would be better if we autogenerated a dbus client for pulseaudio.

For an example of how to do this see:

https://github.com/taffybar/taffybar/blob/master/src/System/Taffybar/DBus/Client/MPRIS2.hs

or anything in 

https://github.com/taffybar/taffybar/blob/master/src/System/Taffybar/DBus/Client

**Comment by u11gh** [comment]
I already checked those files, but I found it much easier to start from
scratch.  That's probably because I have absolutely no experience with
dbus.  (BTW: Progress will slow down as I don't have much time.)

**Comment by IvanMalison** [comment]
> I already checked those files, but I found it much easier to start from scratch.

@u11gh I can make the client for you if you want. It will automatically generate haskell functions to call all of the functions/attach signals/get values of properties that you want. It will be a lot tidier than manually making the function calls yourself.

**Comment by u11gh** [comment]
@IvanMalison

> @u11gh I can make the client for you if you want.

Thanks.  For now I'd like to keep it simple.

Reading and writing the 'Volume' property works already.  What doesn't
work (yet) is listening to the 'VolumeUpdated' signal.  I'll see to it.

**Comment by mt-caret** [comment]
I hacked together a volume widget based on xmobar's approach (https://github.com/jaor/xmobar/blob/master/src/Xmobar/Plugins/Monitors/Volume.hs#L147-L155 ). While it doesn't use the dbus interface mentioned above, it won't be that much work to clean this up into a PR that adds a textVolumeMonitor; is that OK with you?

**Comment by IvanMalison** [comment]
@mt-caret That uses alsa mixer right?

@NickHu added something like this in https://github.com/taffybar/taffybar/pull/203, but I decided to remove it later because I realized that the alsa dependency could prevent e.g. bsd users from being able to install taffybar.

I don't want to add any new non haskell dependencies unless there is a really good reason for it. In this case in particular, there is a way to get the functionality that we want without doing anything like that, and as such, I think that is the approach I want to go with.

**Comment by IvanMalison** [comment]
@mt-caret I believe that at @LeifW and I had discussed making a separate repository that would be used for linux specific taffybar widgets. Let me see if I can find that discussion.

**Comment by mt-caret** [comment]
Aha, so that's why it disappeared; thanks for the info. I guess I'll just keep in my taffybar config.

**Comment by mt-caret** [comment]
Oh, that'll be nice; I would love to merge https://github.com/taffybar/taffybar/pull/65 and https://github.com/taffybar/taffybar/pull/203 and contribute

**Comment by IvanMalison** [comment]
Yeah the aforementioned discussion is in https://github.com/taffybar/taffybar/pull/347#issuecomment-394521321

**Comment by IvanMalison** [comment]
I guess I never made the linux specific taffybar module. I belive that xmobar has cabal feature flags to handle this sort of thing. That would be another option here.

**Comment by mt-caret** [comment]
Feel free to ping me when you make a taffybar-extras repo!

**Comment by IvanMalison** [comment]
https://github.com/taffybar/taffybar-linux

**Comment by IvanMalison** [comment]
literally empty right now so you'll need to make all the project files.

**Comment by mt-caret** [comment]
My Haskell is a bit rusty so I'm not 100% sure how to proceed with setting up the project; is it correct to assume that Nix + stack is preferred?

**Comment by IvanMalison** [comment]
stack/nix is an either thing. both are set up in the taffybar repo. You can use whatever feels comfortable to you (including just raw cabal).

**Comment by mt-caret** [comment]
After using the volume monitor I've made for a while, I realized that it's horribly broken (handling multiple output sources) and I'm not even sure that xmobar's approach is doing the right thing.
I've also running into various stability issues (taffybar suddenly crashing, etc.) so I've decided to move back to xmobar; as such, I'm terribly sorry but I won't be able to work on this.
